src/data/loaders/geojson.ts

```diff
@@ -0,0 +1,198 @@
+type GeoJSONMultiPolygon = {
+  type: "MultiPolygon";
+  coordinates: GeoJSONPosition[][][];
+};
+type GeoJSONGeometry =
+  | GeoJSONPolygon
+  | GeoJSONMultiPolygon
+  | { type: string; coordinates?: unknown };
+
+type GeoJSONFeature = {
+  type: "Feature";
+  geometry: GeoJSONGeometry | null;
+  properties?: Record<string, unknown> | null;
+};
+
+type GeoJSONFeatureCollection = {
+  type: "FeatureCollection";
+  features: GeoJSONFeature[];
+};
+
+type GeoJSONObject =
+  | GeoJSONFeatureCollection
+  | GeoJSONFeature
+  | GeoJSONGeometry
+  | null;
+
+export class GeoJSONShapesDataLoader extends AbstractShapesDataLoader<
+  CompleteGeoJSONShapesDataSource,
+  GeoJSONShapesData
+> {
+  async loadShapes(
+    options: { signal?: AbortSignal } = {},
+  ): Promise<GeoJSONShapesData> {
+    const { signal } = options;
+    signal?.throwIfAborted();
+    const geo = await this._loadGeoJSON({ signal });
+    signal?.throwIfAborted();
+    const multipolygons = this._parseGeoJSONToMultiPolygons(geo);
+    signal?.throwIfAborted();
+    return new GeoJSONShapesData(multipolygons);
+  }
+
+  private async _loadGeoJSON(
+    options: { signal?: AbortSignal } = {},
+  ): Promise<GeoJSONObject> {
+    const { signal } = options;
+    signal?.throwIfAborted();
+    if (this.dataSource.path !== undefined && this.workspace !== null) {
+      const fh = await this.workspace.getFileHandle(this.dataSource.path);
+      signal?.throwIfAborted();
+      const file = await fh.getFile();
+      signal?.throwIfAborted();
+      const text = await file.text();
+      return JSON.parse(text) as GeoJSONObject;
+    }
+    if (this.dataSource.url !== undefined) {
+      const resp = await fetch(this.dataSource.url);
+      if (!resp.ok) {
+        throw new Error(
+          `Failed to fetch GeoJSON: ${resp.status} ${resp.statusText}`,
+        );
+      }
+      const text = await resp.text();
+      return JSON.parse(text) as GeoJSONObject;
+    }
+    if (this.dataSource.path !== undefined) {
+      throw new Error("An open workspace is required to open local-only data.");
+    }
+    throw new Error("A URL or workspace path is required to load data.");
+  }
+
+  private _parseGeoJSONToMultiPolygons(obj: GeoJSONObject): MultiPolygon[] {
+    if (!obj) return [];
+
+    const multiPolygons: MultiPolygon[] = [];
+
+    const pushFromGeometry = (geometry: GeoJSONGeometry | null) => {
+      if (!geometry || !("type" in geometry)) return;
+      const type = geometry.type;
+      if (type === "Polygon") {
+        const coords = (geometry as GeoJSONPolygon).coordinates;
+        const polygon = this._coordsToPolygon(coords);
+        multiPolygons.push({ polygons: [polygon] });
+      } else if (type === "MultiPolygon") {
+        const coords = (geometry as GeoJSONMultiPolygon).coordinates;
+        const polygons = coords.map((polygonCoords) =>
+          this._coordsToPolygon(polygonCoords),
+        );
+        multiPolygons.push({ polygons });
+      }
+    };
```

Comment:
For readability (and performance, in the general case), I would suggest to avoid nested functions